spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,283 @@
+  private static final String TYPE_KEY = "type";
+  private static final String DOT = ".";
+  private static final String DELEGATION_TOKEN_RENEWAL_ENABLE = "delegation.token.renewal.enabled";
+  private static final String SPARK_KERBEROS_KEYTAB = "spark.kerberos.keytab";
+
+  @Override
+  public String serviceName() {
+    return SERVICE_NAME;
+  }
+
+  /**
+   * Builds a HiveConf object for the specified catalog by merging the provided Hadoop configuration
+   * with catalog-specific configurations from the Spark configuration.
+   *
+   * @param sparkConf The SparkConf object containing Spark configurations.
+   * @param hadoopConf The Hadoop Configuration object to be used as the base for the HiveConf.
+   * @param catalogName The name of the catalog.
+   * @return An Optional containing the constructed HiveConf if successful, or an empty Optional if
+   *     an error occurs.
+   */
+  Optional<HiveConf> buildHiveConf(
+      SparkConf sparkConf, Configuration hadoopConf, String catalogName) {
+    String targetPrefix = SPARK_SQL_CATALOG_PREFIX + catalogName + DOT;
+    try {
+      HiveConf hiveConf = new HiveConf(hadoopConf, HiveConf.class);
+      Arrays.stream(sparkConf.getAllWithPrefix(targetPrefix))
+          .forEach(
+              entry -> {
+                if (entry._1.contains(URI_KEY)) {
+                  hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, entry._2);
+                } else {
+                  hiveConf.set(entry._1, entry._2);
+                }
+              });
+      // The `RetryingHiveMetaStoreClient` may block the subsequent token obtaining,
+      // and `obtainDelegationTokens` is scheduled frequently, it's fine to disable
+      // the Hive retry mechanism.
+      hiveConf.set(HiveConf.ConfVars.METASTORE_FASTPATH.varname, "false");
+
+      return Optional.of(hiveConf);
+    } catch (Exception e) {
+      LOG.warn("Fail to create Hive Configuration for catalog {}: {}", catalogName, e.getMessage());
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Extracts the names of Iceberg catalogs from the provided Spark configuration. This method
+   * filters the Spark configuration entries to identify those that correspond to Iceberg catalogs.
+   * It checks if the catalog type matches either `SparkSessionCatalog` or `SparkCatalog` and
+   * collects their names.
+   *
+   * @param sparkConf The SparkConf object containing Spark configurations.
+   * @return A Set of Strings representing the names of Iceberg catalogs.
+   */
+  private Set<String> extractIcebergCatalogNames(SparkConf sparkConf) {
+    return Arrays.stream(sparkConf.getAllWithPrefix(SPARK_SQL_CATALOG_PREFIX))
+        .filter(
+            entry -> {
+              String val = entry._2();
+              return val.contains(SparkSessionCatalog.class.getName())
+                  || val.contains(SparkCatalog.class.getName());
+            })
+        .map(Tuple2::_1)
+        .collect(Collectors.toSet());
```

Comment:
you should filter out illegal catalog names, see SPARK-39079

please also add UT to these method